go/signedexchange/certurl/ocsp.go

```diff
@@ -3,39 +3,61 @@ package certurl
 import (
 	"bytes"
 	"crypto/x509"
+	"encoding/base64"
 	"fmt"
 	"golang.org/x/crypto/ocsp"
 	"io"
 	"io/ioutil"
 	"net/http"
+	"net/url"
 )
 
-func CreateOCSPRequest(certs []*x509.Certificate) (*http.Request, error) {
+func CreateOCSPRequest(certs []*x509.Certificate, preferGET bool) (*http.Request, error) {
 	if len(certs) < 2 {
 		return nil, fmt.Errorf("Could not fetch OCSP response: Issuer certificate not found")
 	}
 	cert := certs[0]
 	if len(cert.OCSPServer) == 0 {
 		return nil, fmt.Errorf("Could not fetch OCSP response: No OCSP responder field")
 	}
-	ocspUrl := cert.OCSPServer[0]
+	ocspUrl, err := url.Parse(cert.OCSPServer[0])
+	if err != nil {
+		return nil, fmt.Errorf("Could not fetch OCSP respones: Invalid OCSP responder URL: %s", err)
+	}
 	issuer := certs[1]
 
 	ocspRequest, err := ocsp.CreateRequest(cert, issuer, &ocsp.RequestOptions{})
 	if err != nil {
 		return nil, err
 	}
-	request, err := http.NewRequest("POST", ocspUrl, bytes.NewReader(ocspRequest))
-	if err != nil {
-		return nil, err
+
+	// Use GET for small requests per https://tools.ietf.org/html/rfc5019#section-5.
+
+	// Use QueryEscape here instead of PathEscape to escape not only '/' but also '+' and '='
+	// to align with the exmaple in https://tools.ietf.org/html/rfc5019#section-5.
+	// Note that the string to be escaped doesn't contain other symbols as it is base64 encoded.
+	getUrl, err := ocspUrl.Parse(url.QueryEscape(base64.StdEncoding.EncodeToString(ocspRequest)))
```

Comment:
Per https://tools.ietf.org/html/rfc2560#appendix-A.1.1, `getUrl` should be constructed as follows:
```
{url}/{url-encoding of base-64 encoding of the DER encoding of the OCSPRequest}
```
`ocspUrl.Parse(...)` may produce different result, for example if `ocspUrl` is `http://example.com/ocsp`.

Reply:
Thanks, fixed it.